code/mob/living/critter/sawfly.dm

```diff
@@ -194,41 +184,68 @@ This file is the critter itself, and all the custom procs it needs in order to f
 			explosion(src, get_turf(src), 0, 0.75, 1.5, 3)
 			qdel(src)
 
-		if(isalive(src)) // if they get EMP'd, they don't *actually* die, so we'll want to fix that
+		if(!isdead(src)) // if they get EMP'd, they don't *actually* die, so we'll want to fix that
 			qdel(src)
 
 	attack_hand(var/mob/user as mob)
-		if (issawflybuddy(user) || (user in src.friends))
-			if (user.a_intent == INTENT_HELP || user.a_intent == INTENT_GRAB)
-				if(isalive(src))
-					src.is_npc = FALSE
+		if (user.a_intent == INTENT_HELP || user.a_intent == INTENT_GRAB)
+			if (issawflybuddy(user) || (user in src.friends))
+				if(!isdead(src))
+					src.ai?.disable()
 					boutput(user, "You collapse [src].")
 					src.foldself()
-		else
-			if(prob(50)&& isalive(src))
-				boutput(user, "<span class='alert' In your attempt to pet [src], you cut yourself on it's blades!</span>")
-
+			else
+				if(prob(50)&& !isdead(src))
+					boutput(user, "<span class='alert'>In your attempt to pet [src], you cut yourself on it's blades!</span>")
 				random_brute_damage(user, 7)
 				take_bleeding_damage(user, null, 7, DAMAGE_CUT, 1)
+		else //harm or shove intent is an attack
+			do_retaliate(user)
 		..()
 
 	Life()
-		if(src.dontdolife) //prevents them from doing much of anything when in grenade form
-			return
 		..()
-		if(prob(8)) communalbeep()
-		if(!isalive(src)) src.set_density(FALSE) //something in the mob life resets density so it has to be below parent
+		if(prob(8) && isturf(src.loc)) communalbeep() //beep only when not in a grenade
+
+
+	seek_target(range) //ai mob critter targetting behaviour - returns a list of acceptable targets
+		if(src.lastattacker && src.retaliate && GET_DIST(src, src.lastattacker) <= range)
+			return list(src.lastattacker)
+		var/targetcount = 0
+		. = list()
+		for (var/mob/living/C in viewers(range, src))
+			if (C.health < -50 || isdead(C))
+				continue
+			if(istype(C, /mob/living/critter/robotic/sawfly))
+				continue
+			if (isintangible(C))
+				continue
+			if(C.mind?.special_role && issawflybuddy(C))
+				if(!(C.weakref in src.friends))
+					boutput(C, "<span class='alert'>[src]'s IFF system silently flags you as an ally! </span>")
+					src.friends += get_weakref(C)
+				continue
+			if(C.job in list( "Head of Security", "Security Officer", "Nanotrasen Security Consultant")) //hopefully this is cheaper than the OR chain I had before
```

Comment:
Should make this a static list on the type, bit faster that way.
```suggestion
			if(C.job in list("Head of Security", "Security Officer", "Nanotrasen Security Consultant")) //hopefully this is cheaper than the OR chain I had before
```
(also space)

Reply:
done :)